habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py

```diff
@@ -1128,92 +1156,189 @@ def set_link_normalized_joint_position(
     """
 
-    set_link_normalized_joint_position(objectA, link_ix, 0)
+    set_link_normalized_joint_position(object_a, link_ix, 0)
 
 
 def open_link(
-    objectA: habitat_sim.physics.ManagedArticulatedObject, link_ix: int
+    object_a: habitat_sim.physics.ManagedArticulatedObject, link_ix: int
 ) -> None:
     """
     Set a link to the "open" state. Sets the joint position to the maximum joint limit.
 
     TODO: does not do any collision checking to validate the state or move any other objects which may be contained in or supported by this link.
 
-    :param objectA: The parent ArticulatedObject of the link to check.
+    :param object_a: The parent ArticulatedObject of the link to check.
     :param link_ix: The index of the link within the parent object. Not the link's object_id.
     """
 
-    set_link_normalized_joint_position(objectA, link_ix, 1.0)
+    set_link_normalized_joint_position(object_a, link_ix, 1.0)
+
+
+def bb_next_to(
+    bb_a: mn.Range3D,
+    bb_b: mn.Range3D,
+    transform_a: mn.Matrix4 = None,
+    transform_b: mn.Matrix4 = None,
+    vertical_threshold=0.1,
+    l2_threshold=0.3,
+) -> bool:
+    """
+    Check whether or not two bounding boxes should be considered "next to" one another.
+    Concretely, consists of two checks:
+     1. height difference between the lowest points on the two objects to check that they are approximately resting on the same surface.
+     2. regularized L2 distance between object centers. Regularized in this case means displacement vector is truncted by each object's heuristic size.
+
+    :param bb_a: local bounding box of one object
+    :param bb_b: local bounding box of another object
+    :param transform_a: local to global transform for the first object. Default is identity.
+    :param transform_b: local to global transform for the second object. Default is identity.
+    :param vertical_threshold: vertical distance allowed between objects' lowest points.
+    :param l2_threshold: regularized L2 distance allow between the objects' centers.
+
+    :return: Whether or not the objects are heuristically "next to" one another.
+    """
+
+    if transform_a is None:
+        transform_a = mn.Matrix4.identity_init()
+    if transform_b is None:
+        transform_b = mn.Matrix4.identity_init()
+
+    keypoints_a = get_global_keypoints_from_bb(bb_a, transform_a)
+    keypoints_b = get_global_keypoints_from_bb(bb_b, transform_b)
+
+    lowest_height_a = min([p[1] for p in keypoints_a])
+    lowest_height_b = min([p[1] for p in keypoints_b])
+
+    if abs(lowest_height_a - lowest_height_b) > vertical_threshold:
+        return False
+
+    if (
+        size_regularized_bb_distance(bb_a, bb_b, transform_a, transform_b)
+        > l2_threshold
+    ):
+        return False
+
+    return True
+
+
+def obj_next_to(
+    sim: habitat_sim.Simulator,
+    object_id_a: int,
+    object_id_b: int,
+    vertical_threshold=0.1,
+    l2_threshold=0.5,
```

Comment:
I might look into setting the `l2_threshold` relative to object size. These utilities support that well 👍

Reply:
Yeah, I already regularize the distance by object size, but it is true that our semantic notion of "nearby" or "next to" is vastly different for different object scales. I wonder if there is some nice "golden number" or ratio to capture this as a perceptual phenomena. :thinking: